### PR TITLE
Delegate CLaaS training to upstream veRL workers

### PR DESCRIPTION
**Pull request by kfallah**
CLaaS delegates training to public veRL 0.9.0 `TrainingWorker` and its FSDP engine. Upstream code owns model/LoRA construction, forward/backward microbatching, optimizer updates, the scheduler, and native checkpoint save/load. CLaaS provides a full-logit output extension, the SDPO/REINFORCE/hybrid objective callback, and EMA teacher updates. There is no independent CLaaS optimizer loop or CPU training fallback.

Portable jobs retain original prompt/response token IDs, behavior log probabilities, sampling controls, and model/tokenizer/policy identities. One BF16 CUDA rank runs actor and separate forward-only teacher workers, with upstream teacher CPU offload. Full-vocabulary SDPO uses feedback-conditioned teacher distributions; hybrid requires both text and scalar signals. Model and tokenizer snapshots are pinned to immutable Hugging Face revisions. The original dense Qwen3.5 wrapper uses native PEFT adapter names.

Native actor checkpoints include LoRA parameters, optimizer, scheduler, and RNG state; native teacher state and serving adapter exports are included in a strict schema-2, hash-bound receipt. Resume reads a private verified snapshot. Atomic publication, lineage-aware freshness, duplicate-update rejection, scope checks, bounded subprocess cleanup, diagnostic tails, and restricted worker environments remain intact. Older custom-trainer checkpoints are deliberately rejected rather than silently converted. The public training stack is isolated to the optional `claas-verl` extra, which the explicit development extra includes for upstream contract tests; ordinary installation and gateway imports stay independent.

Implementation is grouped under `exp/optimize/claas/backends/verl/` with explicit worker, engine, inputs, objective, and state modules. Shared checkpoint receipts and subprocess transport remain at the backend root. Worker launch uses `python -m exp.optimize.claas.backends.verl.worker`; there are no compatibility modules.

Validation on this head:
- The refreshed capture base includes current main and the bounded catalog helper extraction, with flagship 0.7.84 and native extension 0.3.75. The native extension was rebuilt in the training environment.
- Whole-repository Ruff, format, and ty pass. The focused training, catalog, import-boundary, and layout suite passes 120 tests, with one explicit CUDA integration skip.
- CPU checks cover numerical objectives, exact nested-token alignment through upstream input/output APIs, configuration and original Qwen LoRA naming, native checkpoint receipt requirements, source mutation, publication failures, subprocess lifecycle/security contracts, and the actual renamed module entrypoint. They do not prove an upstream CUDA optimizer update.
- The opt-in CUDA test creates tiny local model/tokenizer files, performs two real upstream updates around native checkpoint resume, checks nonempty Adam step-2 state and resumed teacher EMA, and verifies serving export. It remains unrun because this verification host has no CUDA/NCCL runtime. Full Qwen GPU training/serving, paid provider calls, and Modal GPU execution are also unrun.
- Every current-head CI check passes, Greptile reports an actual 5/5 on `9ebf0ee904a58311ad0101007b0bf0b88582c548`, and no review threads remain unresolved. One mock-request latency job passed on bounded rerun after the identical local CLI completed without failures; no code change was needed. The development-extra security advisory was rebutted with the explicit opt-in, locked dependencies, read-only CI permissions, no provider credentials, and separately gated GPU execution, then dismissed by the security scan.

Run the GPU gate only on explicitly authorized hardware:
```bash
CLAAS_RUN_CUDA_INTEGRATION=1 CUDA_VISIBLE_DEVICES=0 uv run --extra dev --extra claas-verl pytest -q exp/optimize/claas/backends/verl/worker_test.py::test_cuda_upstream_verl_update_native_resume_and_export
```

Stack base: `codex/claas-01-capture`. Scenario/environment generation, serving, evaluation, promotion, and Modal transport remain separate stack layers. No private veRL fork is published. Do not merge.

### CODE REVIEW

**Comment by kfallah** [comment]
@greptileai review

**Comment by kfallah** [comment]
@greptileai review

**Comment by kfallah** [comment]
@greptileai review

**Comment by kfallah** [comment]
@greptileai review

**Comment by kfallah** [comment]
@greptileai review

**Comment by kfallah** [comment]
@greptileai review

File: exp/optimize/claas/backends/subprocess.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                checkpoint_root=str(self._backend._root),
+                resume_checkpoint=self._checkpoint,
+            )
+            with tempfile.TemporaryDirectory(prefix="claas-job-") as directory:
+                root = Path(directory)
+                job_path, result_path = root / "job.json", root / "result.json"
+                job_path.write_text(job.model_dump_json())
+                job_path.chmod(0o600)
+                environment = dict(os.environ)
+                environment["CUDA_VISIBLE_DEVICES"] = self._backend._device
```
[the remark this answers is not in the document]
Comment (kfallah) [reply exp/optimize/claas/backends/subprocess.py]: Fixed in 4b2f11cb and included in b65bb87d. The child now receives an explicit non-secret runtime allowlist, VERL_PLUGINS=none, and HF_HUB_DISABLE_IMPLICIT_TOKEN=1. Optional model access requires a dedicated model_access_token constructor argument. The subprocess regression test installs gateway/cloud credential canaries and verifies that none reach the worker environment. Full gates and all current CI checks pass.

**Comment by kfallah** [comment]
@greptileai review

**Comment by kfallah** [comment]
@greptileai review

Please review current head 77d363756da58823255ba0ee0d92f11cbd39b9d1. The preceding review check reported cancellation; the displayed 5/5 summary still refers to the old head.

**Comment by kfallah** [comment]
@greptileai Please review current head 0ac1606a. This groups veRL integration in its backend package, updates the subprocess module path and narrow dependency allowlist, and fixes helper/ownership docstrings. Whole-repository Ruff, format and ty pass; 75 focused tests pass with one explicit CUDA skip.

**Comment by kfallah** [comment]
@greptileai Please review final package-cleanup head 7fb82e55. The follow-up moves ordinary test imports to module scope per AGENTS.md. Whole-repository Ruff, format and ty pass; 75 focused tests passed after the package move and all 21 affected tests pass after import cleanup. Worker execution through the new module path also passes --help without allocating compute.

**Comment by kfallah** [comment]
@greptileai Review stable head c0779571. Final audit adds a real verl.worker --help regression check and documents the one deferred test-fixture import needed to break its actual cycle. Whole-repository Ruff, format and ty pass; all 23 affected tests pass. Package structure is now frozen pending CI and review.

**Comment by kfallah** [comment]
@greptileai review

File: pyproject.toml [hunk 2]
```diff
@@ -45,6 +48,8 @@ dev = [
     "ruff>=0.5",
     "ty>=0.0.1a1",
     "experiential[sft]",
+    # CPU tests verify objective and worker contracts; real CUDA execution is explicitly opt-in.
```
[the remark this answers is not in the document]
Comment (kfallah) [reply pyproject.toml:51]: This is an intentional developer-test dependency, not an untrusted-input or runtime privilege boundary. The ordinary experiential dependency list does not contain veRL, Torch, Transformers, or PEFT; users opt into claas-verl explicitly. The dev extra is itself an explicit opt-in, and already installs optional SFT SDK dependencies so whole-repository tests can exercise real published integrations. Its adjacent comment states why claas-verl is included: CPU tests verify objective and worker contracts.

Those tests exercise public upstream worker/configuration/output APIs, exact token alignment, objective numerics, checkpoint contracts, and actual LoRA construction. Removing the dependency from dev would silently skip that required regression coverage. GPU execution remains separately gated in worker_test.py:128-130 by both CLAAS_RUN_CUDA_INTEGRATION=1 and an available authorized CUDA device. Installing or importing these dependencies does not launch a GPU, Modal function, or paid provider call.

The committed uv.lock records exact resolved versions and artifact hashes, including verl==0.9.0. The PR gate and latency workflow use contents: read permissions, local/no-live tests, and no provider credentials. The latency benchmark dispatches only to its own loopback mock. The finding identifies no malicious dependency artifact, secret exposure, elevated permission, or caller-controlled package selection. General dependency supply-chain policy is valuable, but moving the same pinned package into another CPU test job would not close a demonstrated security boundary here. Please dismiss this advisory as a test-environment policy preference; we retain the real upstream CPU coverage and the explicit CUDA opt-in.

**Comment by kfallah** [comment]
Superseded by the independent asynchronous learning stack #1018 through #1024. The replacement removes mandatory production-gateway coupling and separates finite burst training, resident full runs, and thin infrastructure hosting. This branch is preserved; no merge or branch deletion is performed.
